tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,438 @@
+
+      subgraph_spec = _handle_nodes_from_other_layers(subgraph_spec,
+                                                      execution_specs,
+                                                      graph,
+                                                      node_name_to_node_def)
+      execution_specs.append(subgraph_spec)
+      previous_layers_visited = previous_layers_visited.union(
+          subgraph_spec.body_node_names)
+
+    # Handle one remote op layer
+    remote_op_specs = _partition_one_remote_op_layer(remote_ops_one_layer,
+                                                     node_name_to_input_names)
+    execution_specs.extend(remote_op_specs)
+
+  # Handle the last subgraph layer
+  output_node_names = set(graph_outputs)
+  subgraph_spec = _partition_one_subgraph_layer(previous_layers_visited,
+                                                graph_def,
+                                                graph,
+                                                output_node_names,
+                                                node_name_to_node_def,
+                                                node_name_to_input_names)
+
+  subgraph_spec = _handle_nodes_from_other_layers(subgraph_spec,
+                                                  execution_specs,
+                                                  graph,
+                                                  node_name_to_node_def)
+  execution_specs.append(subgraph_spec)
+  previous_layers_visited = previous_layers_visited.union(
+      subgraph_spec.body_node_names)
+
+  return execution_specs
+
+
+def _get_subgraph_layer_output_node_names(remote_ops_one_layer,
+                                          node_name_to_node_def,
+                                          node_name_to_input_names):
+  """Get the output names of a subgraph layer.
+
+  Note that they are equivalant to the input names of the succeeding
+    remote op layer.
+  """
+  output_node_names = set([])
+
+  for remote_op in remote_ops_one_layer:
+    for input_node_name in node_name_to_input_names[remote_op]:
+      input_node = node_name_to_node_def[input_node_name]
+
+      # Assumption: graph inputs (placeholders) are loaded before
+      #             executing a graph.
+      if not _is_placeholder_op(input_node):
+        output_node_names.add(input_node_name)
+
+  return output_node_names
+
+
+def _partition_one_subgraph_layer(previous_layers_visited,
+                                  graph_def,
+                                  graph,
+                                  output_names,
+                                  node_name_to_node_def,
+                                  node_name_to_input_names):
+  """Perform a modified BFS for graph partitioning.
+
+  Expand from the outputs, until one of the stopping condition: remote op,
+    placeholder, visited before in this layer, or visited before in the
+    previous layers.
```

Comment:
This function is complicated enough to deserve a longer explanation of the algorithm. Think:

What does "modified" qualifier for BFS mean and how does it work?
Elaborate on the "Expand from the outputs".
What does a "layer" mean (this term is not described anywhere AFAICT)?

Reply:
Added a more detailed docstring to walk through the algorithm.